### PR TITLE
Fixes #7874 - Containers API

### PR DESCRIPTION
**Pull request by eLobato**
This is still being developed, wait until it's ready to be merged. This PR is only so that collaboration with @mbacovsky gets a bit easier.

### CODE REVIEW

**Comment by eLobato** [comment]
Supports all actions but create/power now

**Comment by eLobato** [comment]
Added power action

**Comment by eLobato** [comment]
@mbacovsky the create action here doesn't work yet, but you should get the idea more or less of what parameters to use in hammer, please let me know any suggestions

**Comment by eLobato** [comment]
@mbacovsky This should be done, except for the fact I would prefer this to be in it's own foreman_docker namespace, any ideas @daviddavis ?

**Comment by mbacovsky** [comment]
@eLobato thanks a lot, updated, testing.
One thing worth improving would be extending the rabl templates to contain full `container` attributes, so that the CLI and UI match in the amount of information given about an `container`. I'll add patch to this PR.

**Comment by daviddavis** [comment]
@eLobato it should be possible. We do it in Katello. Might want to look at Katello's api controllers/routes.

File: app/controllers/api/v2/containers_controller.rb [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+      resource_description do
+        resource_id 'containers'
+        api_version 'v2'
+        api_base_url '/api/v2'
+      end
+
+      api :GET, '/containers/', N_('List all containers')
+      api :GET, '/compute_resources/:compute_resource_id/containers/',
+          N_('List all containers in a compute resource')
+      param :compute_resource_id, :identifier
```
Comment (daviddavis) [inline app/controllers/api/v2/containers_controller.rb]: Where is this being used?

File: app/controllers/api/v2/containers_controller.rb [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+                                                 :cpu_shares, :memory],
+                              :environment   => [:tty, :attach_stdin, :attach_stdout,
+                                                 :attach_stderr] }
+
+        wizard_state = DockerContainerWizardState.create
+        wizard_properties.each do |step, properties|
+          property_values = {}
+          properties.each do |property|
+            property_values[:"#{property}"] = params[:container][:"#{property}"]
+          end
```
Comment (daviddavis) [inline app/controllers/api/v2/containers_controller.rb]:
```ruby
property_values = properties.each_with_object({}) |property, values|
  property_values[:"#{property}"] = params[:container][:"#{property}"]
end
```

File: app/controllers/api/v2/containers_controller.rb [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+          wizard_state.send(:"create_#{step}", property_values)
+        end
+
+        if params[:container][:environment_variables].present?
+          wizard_state.environment.environment_variables =
+            params[:container][:environment_variables]
+        end
+        wizard_state.save
+        wizard_state
```
Comment (daviddavis) [inline app/controllers/api/v2/containers_controller.rb]:
```ruby
wizard_state.tap(&:save)
```

**Comment by parthaa** [comment]
@daviddavis @eLobato Is this supposed to handle Creation of both Katello and Foreman docker containers. Asking because katello container create would require the following elements
1) SmartProxy or Capsule id
2) Org Id
3) Lifecycle Environment Id
4) Content View Id
5) Repository Id
6) Tag

**Comment by parthaa** [comment]
@daviddavis we might need to add some logic that says if -> katello is set to true then check for these katello attributes..

**Comment by eLobato** [comment]
@parthaa Not yet ready I'd say. Should we include that in another PR?

File: app/controllers/api/v2/containers_controller.rb (hunk 1 above)
Comment (eLobato) [inline app/controllers/api/v2/containers_controller.rb]: Nowhere actually, I should add the routes so that compute_resources picks this up.

**Comment by daviddavis** [comment]
@eLobato @parthaa a separate PR works for me.

File: app/controllers/api/v2/containers_controller.rb [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+          end
+          wizard_state.send(:"create_#{step}", property_values)
+        end
+
+        if params[:container][:environment_variables].present?
+          wizard_state.environment.environment_variables =
+            params[:container][:environment_variables]
+        end
+        wizard_state.tap(&:save)
+        wizard_state
```
Comment (daviddavis) [inline app/controllers/api/v2/containers_controller.rb]: Remove this line.

**Comment by daviddavis** [comment]
I'd say this is good to go with the understanding that we'll tackle partha's items separately.

**Comment by mbacovsky** [comment]
For me `index' command does not work for some reason in API (UI does). It worked before, but may be some misconfiguration

**Comment by eLobato** [comment]
@mbacovsky I've added some fixes, including namespacing the routes. You should be able to access containers at `docker/api/v2/compute_resources/:compute_resource_id/containers` and `docker/api/v2/containers/`

@daviddavis @parthaa @witlessbird :ok: to merge this without namespacing the code? Just so it makes it for 1.2.0 (to be released tomorrow).

**Comment by daviddavis** [comment]
:+1: from me. Tested and it worked.

Created [an issue from @parthaa's comment](http://projects.theforeman.org/issues/9518).

**Comment by eLobato** [comment]
Merged as 5274f6bc4e53e8e38c4b111c268c039a87102f85, thanks @daviddavis @mbacovsky @parthaa

**Pull request closed without merging**